rs/ingress_manager/src/ingress_handler.rs

```diff
@@ -58,6 +58,11 @@ impl<T: IngressPool> ChangeSetProducer<T> for IngressManager {
             .map(|artifact| {
                 let ingress_object = &artifact.message;
 
+                // If the ingress pool is full, discard the message
+                if pool.exceeds_limit(&ingress_object.peer_id) {
+                    return RemoveFromUnvalidated(IngressMessageId::from(ingress_object));
```

Comment:
I am not sure this is correct. If you do it like this won't it delete all unvalidated artifacts from the peer that exceeds the limit because the counter is not updated immediately ?

Reply:
as commented, please just explain why this will work and explain when it will be fixed

Reply:
Can you please take a look if the comment makes sense?

Reply:
The comment explains the case for discarding sufficiently, imho.

However, we also have a similar problem when the pool doesn't exceed the limit. I.e., this code may insert several messages, even though a single one has brought the pool over one of its limits.

Is my assumption correct that there can't be more than 1 concurrent executions of on_state_change? In this case we should be fine because of the boundedness of the pools.

Reply:
> Is my assumption correct that there can't be more than 1 concurrent executions of on_state_change? In this case we should be fine because of the boundedness of the pools.

That's my understanding